substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/thread/PlatformThreads.java

```diff
@@ -251,6 +251,18 @@ public static long getThreadCpuTime(long javaThreadId, boolean includeSystemTime
         }
     }
 
+    @Uninterruptible(reason = "Thread locks/holds the THREAD_MUTEX.")
+    public static void iterateIsolateThreads(IsolateThreadConsumer consumer) {
```

Comment:
The `IsolateThreadConsumer` has been added because `VMThreads.lockThreadMutexInNativeCode()` method has package access. For example `PlatformThreads.getThreadCpuTime(...)` is in the same package as `VMThreads` so it is able to call `VMThreads.lockThreadMutexInNativeCode()` to iterate over threads.
https://github.com/oracle/graal/blob/114067fc41d97e6c07f6de9bd745196d6f967ae4/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/thread/PlatformThreads.java#L241

I updated the code to call VMOperation to iterate over threads to emit ThreadCPULoad event.
I tried to check `JfrEvent.ThreadCPULoad.shouldEmit()` first to not call VMOperation in case it is not necessary but the method which uses `JfrEvent.ThreadCPULoad.shouldEmit()` needs to be `@Uninterruptible` and I do not see a way to call VMOperation from uninterruptible method because `JavaVMOperation.enqueue()` is not  uninterruptible.

Reply:
You could try annotating `ThreadCpuLoadEvent.emitEvents()` with `@Uninterruptible(reason = "Check should emit as a minor optimization.", calleeMustBe = false)`. I think this should be ok because you check `JfrEvent.ThreadCPULoad.shouldEmit()` again from uninterruptible code once the VM operation has started.

Reply:
What I can see the first case requires that VMOperation has been created outside `@Uninterruptible` method.
Something like:
```
    public static void emitEvents() {
        emitEvents(new EmitThreadCPULoadEventsVMOperation());
    }

    @Uninterruptible(reason = "Check should emit as a minor optimization.", calleeMustBe = false)
    public static void emitEvents(EmitThreadCPULoadEventsVMOperation vmOp) {
        if (JfrEvent.ThreadCPULoad.shouldEmit()) {
            vmOp.enqueue();
        }
    }
```

I applied the second one which allows to check `shouldEmit()` before the VMOperation is created.
```
    @Uninterruptible(reason = "Check should emit in emit events as a minor optimization.")
    private static boolean shouldEmit() {
        return JfrEvent.ThreadCPULoad.shouldEmit();
    }

    public static void emitEvents() {
        if (shouldEmit()) {
            EmitThreadCPULoadEventsVMOperation vmOp = new EmitThreadCPULoadEventsVMOperation();
            vmOp.enqueue();
        }
    }
```